File: frontend/src/types/transaction.ts
import { CategoryType } from './category';

export interface Transaction {
  id: string;
  user_id: string;
  category_id: string;
  category_name: string;
  category_icon: string;
  category_color: string;
  type: CategoryType;
  amount: number;
  title: string;
  description?: string | null;
  transaction_date: string;
  created_at: string;
  updated_at?: string;
}

export interface TransactionPagination {
  page: number;
  limit: number;
  total_count: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface TransactionListResponse {
  transactions: Transaction[];
  pagination: TransactionPagination;
}

export interface TransactionFilters {
  type?: CategoryType | '';
  category_id?: string;
  start_date?: string;
  end_date?: string;
  search?: string;
  sort_by?: 'transaction_date' | 'amount' | 'title' | 'created_at';
  sort_order?: 'asc' | 'desc';
  page?: number;
  limit?: number;
}

export interface CreateTransactionPayload {
  category_id: string;
  type: CategoryType;
  amount: number;
  title: string;
  description?: string;
  transaction_date: string;
}

export interface UpdateTransactionPayload {
  category_id?: string;
  type?: CategoryType;
  amount?: number;
  title?: string;
  description?: string;
  transaction_date?: string;
}

File: backend/app/models/transaction.py
import uuid
from datetime import datetime, timezone
from app.extensions import db

class Transaction(db.Model):
    __tablename__ = "transactions"

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = db.Column(db.String(36), db.ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    category_id = db.Column(db.String(36), db.ForeignKey("categories.id", ondelete="RESTRICT"), nullable=False, index=True)
    type = db.Column(db.String(10), nullable=False, index=True)  # 'income' or 'expense'
    amount = db.Column(db.Numeric(12, 2), nullable=False)
    title = db.Column(db.String(150), nullable=False)
    description = db.Column(db.Text, nullable=True)
    transaction_date = db.Column(db.Date, nullable=False, index=True)
    created_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    # Relationships
    user = db.relationship("User", back_populates="transactions")
    category = db.relationship("Category", back_populates="transactions")

    def to_dict(self):
        return {
            "id": self.id,
            "user_id": self.user_id,
            "category_id": self.category_id,
            "category_name": self.category.name if self.category else "Uncategorized",
            "category_icon": self.category.icon if self.category else "Tag",
            "category_color": self.category.color if self.category else "#64748B",
            "type": self.type,
            "amount": float(self.amount),
            "title": self.title,
            "description": self.description,
            "transaction_date": self.transaction_date.isoformat() if self.transaction_date else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }
